lib/com/codeblock.tsx

```diff
@@ -63,45 +81,124 @@ export class CodeBlock {
+        if (ctx.node.parent && ctx.node.parent.hasComponent(Document))
+          return false;
         return true;
       },
       action: (ctx: Context) => {
         const com = new CodeBlock();
-        ctx.node.addComponent(com);
-        ctx.node.changed();
-        workbench.workspace.setExpanded(ctx.path.head, ctx.path.node, true);
-      }
+        if (ctx?.node) {
+          ctx.node.addComponent(com);
+          ctx.node.changed();
+          workbench.workspace.setExpanded(
+            ctx.path.head,
+            ctx.path.node,
+            true
+          );
+        }
+      },
     });
   }
-
 }
 
-
 const CodeEditor = {
-  oncreate({dom, attrs: {path}}) {
+  oncreate(vnode) {
+    const {
+      dom,
+      attrs: { path },
+    } = vnode;
     const snippet = path.node.getComponent(CodeBlock);
+    //@ts-ignore
     dom.jarEditor = new window.CodeJar(dom, (editor) => {
       // highlight.js does not trim old tags,
       // let's do it by this hack.
       editor.textContent = editor.textContent;
+      //@ts-ignore
       window.hljs.highlightBlock(editor);
-      snippet.language = window.hljs.highlightAuto(editor.textContent).language || "";
+      snippet.language =
+        //@ts-ignore
+        window.hljs.highlightAuto(editor.textContent).language || "";
     });
     dom.jarEditor.updateCode(snippet.code);
-    dom.jarEditor.onUpdate(code => {
+    dom.jarEditor.onUpdate((code) => {
       snippet.code = code;
       path.node.changed();
     });
   },
 
-  view({attrs: {workbench, path}}) {
+  view({ attrs: { workbench, path } }) {
     // this cancels the keydown on the outline node
     // so you can use arrow keys normally
     const onkeydown = (e) => e.stopPropagation();
-    
-    return (
-      <div class="code-editor" onkeydown={onkeydown}></div>
-    )
+
+    return <div class="code-editor" onkeydown={onkeydown}></div>;
+  },
+};
+
+class Output {
+  constructor(vnode) {
+    // Initialize output with m.prop() to preserve state across re-renders
+    this.output = vnode.attrs.output || "";
+    this.updateOutput = vnode.attrs.updateOutput || (() => {});
   }
-}
\ No newline at end of file
+
+  oncreate(vnode) {
+    const {
+      dom,
+      attrs: { path },
+    } = vnode;
+
+    const snippet = path.node.getComponent(CodeBlock);
+
+    // Apply syntax highlighting
+    window.hljs.highlightBlock(dom);
+
+    // Find the button element and add event listener
+    const button = dom.querySelector("button");
+    if (button) {
+      button.addEventListener("click", async () => {
+        console.log("Button clicked");
+        console.log(snippet);
+
+        try {
+          const res = await defaultExecutor.execute(snippet.code, {
+            language: snippet.language,
+          });
+          console.log("Execution result:", res);
+
+          // Update output using m.prop to ensure it's persistent across re-renders
+          this.output = res; // Call m.prop with the new value
```

Comment:
Although we *can* put output state on the component, we aren't really going to do much with it other than display it. So a better place to put this would be in the output component's state, which is persisted across redraws and is a reasonable place to put stuff from other systems into a component to display.

Reply:
Sorry, my language was a little ambiguous and I didn't see exactly what was happening. Let me start over:

First, although putting fields on a *Mithril component* class technically works as Mithril component state, I would prefer to directly use the `state` on the `vnode`. In fact, there are several ways to define a Mithril component but after trying a bunch I've standardized on the non-class approach. So like `CodeEditor`. 

And then because this output state isn't really needed by the *Treehouse component*, we don't need to have fields on `CodeBlock`. I'm just now realizing I need to be more specific because a Treehouse component is **not** a Mithril component.